libgraphics/source/types/TransformArray.cxx

```diff
@@ -0,0 +1,120 @@
+#include <cpplogger/Logger.hpp>
+
+#include <pivot/graphics/types/TransformArray.hxx>
+
+using Transform = pivot::graphics::Transform;
+
+namespace pivot::graphics
+{
+
+void TransformArray::setValueForEntity(Entity entity, std::optional<ecs::data::Value> value)
+{
+    if (value.has_value()) {
+        if (entity >= m_components.size()) {
+            m_components.resize(entity + 1);
+            m_component_exist.resize(entity + 1, false);
+            m_reverse_root.resize(entity + 1, std::set<Entity>{});
+        }
+
+        Transform newTransform = this->parseValue(value.value());
+        if (newTransform.root.is_empty()) {
+            this->removeRoot(entity);
+        } else {
+            this->setRoot(entity, newTransform.root.ref);
+        }
+
+        Transform &entityTransform = m_components.at(entity);
+        entityTransform.position = newTransform.position;
+        entityTransform.rotation = newTransform.rotation;
+        entityTransform.scale = newTransform.scale;
+        m_component_exist.at(entity) = true;
+    } else {
+        this->removeTransform(entity);
+    }
+}
+
+void TransformArray::setRoot(Entity entity, Entity root)
+{
+    Transform &transform = this->m_components.at(entity);
+
+    // Handle the old root only if the entity exist. Otherwise the root value is
+    // just some remaining value from a previous entity
+    if (this->entityHasValue(entity)) {
+        // If the old root is the same as the new root, no need to do anything
+        if (!transform.root.is_empty() && transform.root.ref == root) { return; }
+
+        // Otherwise, remove the old root
+        this->removeRoot(entity);
+    }
+
+    // Check that the new root corresponds to an entity with a Transform component
+    if (!this->entityHasValue(root)) {
+        logger.warn() << "Cannot set transform root to entity without transform";
+        transform.root = EntityRef::empty();
+        return;
+    }
+
+    // Check that the new root is not the entity itself
+    if (entity == root) {
+        logger.warn() << "Cannot set transform root to self";
+        transform.root = EntityRef::empty();
+        return;
+    }
+
+    const Transform &rootTransform = this->m_components.at(root);
+
+    // If the new root also has a root, the depth is larger than 1
+    if (!rootTransform.root.is_empty()) { throw RootDepthExceeded(entity); }
+
+    // If the entity is the root of other entities, the depth is larger than 1
+    if (!m_reverse_root.at(entity).empty()) { throw RootDepthExceeded(*m_reverse_root.at(entity).begin()); }
```

Comment:
I know we've talked about this, but I still don't think that imposing a limit of only one root is a good thing, but whatever.

On the other hand, preventing one entity from having more than one child serves no purpose, please remove it

Reply:
Nooo that's also the root check.

```
// First check
other ---> root --error--> entity

// Second check
root --error--> entity --> other
```

Reply:
Oh ! You're right !